### PR TITLE
feat(openresty): add option to build boringssl 

### PR DESCRIPTION
**Pull request by mayocream**
# Summary

Add boringssl support to docker build scripts.
related: https://github.com/Kong/kong-ee/pull/3313

### CODE REVIEW

**Comment by curiositycasualty** [comment]
cc @fffonion

**Comment by fffonion** [comment]
@mayocream let's pair on this today

File: Jenkinsfile [hunk 1]
```diff
@@ -91,13 +91,30 @@ pipeline {
...
+                    environment {
+                        PATH = "/home/ubuntu/bin/:${env.PATH}"
+                        GITHUB_SSH_KEY = credentials('github_bot_ssh_key')
+                    }
+                    steps {
+                        sh 'echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin || true'
+                        sh 'while /bin/bash -c "ps aux | grep [a]pt-get"; do sleep 5; done'
+                        sh 'curl https://raw.githubusercontent.com/Kong/kong/master/scripts/setup-ci.sh | bash'
+                        sh 'git clone --recursive --single-branch --branch ${KONG_SOURCE} git@github.com:Kong/kong-ee.git ${KONG_SOURCE_LOCATION}'
+                        sh 'make CACHE=false PACKAGE_TYPE=deb RESTY_IMAGE_BASE=ubuntu RESTY_IMAGE_TAG=20.04 SSL_PROVIDER=boringssl package-kong test cleanup'
```
Comment (hutchic) [inline Jenkinsfile:115]: I think it's ok to leave this as a task for future us but we should resolve `CACHE=false` at some point

File: build-kong.sh [hunk 2]
```diff
@@ -20,16 +20,20 @@ export LUA_PATH="/usr/local/share/lua/5.1/?.lua;/usr/local/openresty/luajit/shar
...
 
 /usr/local/bin/luarocks --version
-/usr/local/kong/bin/openssl version
-/usr/local/openresty/bin/openresty -v
+/usr/local/kong/bin/openssl version || true
+ldd /usr/local/openresty/nginx/sbin/nginx || true
+strings /usr/local/openresty/nginx/sbin/nginx | grep rpath || true
+strings /usr/local/openresty/bin/openresty | grep rpath || true
+find /usr/local/kong/lib/ || true
+/usr/local/openresty/bin/openresty -V || true
```
Comment (hutchic) [inline build-kong.sh:28]: are these lines more informational now than testing anything?

File: openresty-build-tools/kong-ngx-build [hunk 3]
```diff
@@ -600,41 +618,89 @@ main() {
...
-  if [ -n "$BORINGSSL_VER" ]; then
+  if [ "$BORINGSSL_VER" != 0 -a "$SSL_PROVIDER" = "boringssl" ]; then
     if [ ! -d "$BORINGSSL_DOWNLOAD/.openssl/lib" ]; then
       notice "Building BoringSSL..."
 
+      arch=$(uname -m)
+      if [ "$(arch)" != "x86_64" ]; then
+        echo "Error: BoringSSL builds only supported in x86_64; got $arch" >&2
+        exit 1
+      fi
```
Comment (hutchic) [inline openresty-build-tools/kong-ngx-build:629]: dissapointed but acceptable for now

**Review by hutchic — APPROVED** [review]

File: Jenkinsfile [hunk 4]
```diff
@@ -91,13 +91,30 @@ pipeline {
...
                 }
+                stage('Kong Enterprise BoringSSL') {
+                    agent {
+                        node {
+                            label 'bionic'
+                        }
+                    }
+                    environment {
+                        PATH = "/home/ubuntu/bin/:${env.PATH}"
+                        GITHUB_SSH_KEY = credentials('github_bot_ssh_key')
```
Comment (hutchic) [inline Jenkinsfile:108]:
```suggestion
                        GITHUB_SSH_KEY = credentials('github_bot_ssh_key')
                        DOCKER_REPOSITORY = "kong/kong-build-tools-private"
```

**Review by hutchic — CHANGES_REQUESTED** [review]
changing to requesting a change -- only the `DOCKER_REPOSITORY` environment variable. I know `CACHE` is set to false so it shouldn't matter but best be sure we don't accidentally push enterprise code to a public docker hub repository

File: Jenkinsfile (hunk 4 above)
Comment (hutchic) [reply Jenkinsfile:108]: nevermind github diff confused me we're covered already :)

**Review by hutchic — APPROVED** [review]

File: openresty-build-tools/kong-ngx-build (hunk 3 above)
Comment (gszr) [reply openresty-build-tools/kong-ngx-build:629]: I am unsure the boringssl integrity checks work on arm64 boxes. On my m1, it did not. We will need to circle back.

File: build-kong.sh (hunk 2 above)
Comment (gszr) [reply build-kong.sh:28]: yeah, debugging purposes

**Comment by fffonion** [comment]
I have verified @gszr 's work on ubuntu20.04 build, it's working fine and I can enable the `fips=on` switch. Couple of things left:
- [ ] one of RHEL support
- [ ] add the `~fips` marker in version (kong version, package version filename etc), as the 4 digit version revert is not merged yet, hold this for now.

**Comment by aboudreault** [comment]
I have created tickets and assigned GS for the remaining tasks.

**Pull request merged**